Support default values on NamedTuple fields

ansley opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#54682**

Differential Revision: [D27327241](https://our.internmc.facebook.com/intern/diff/D27327241)

Commit:
Support default values on NamedTuple fields

[ghstack-poisoned]

gmagogsfm commented on test/test_jit.py:
```diff
@@ -15632,7 +15649,7 @@ def fn(x=None):
         self.checkScript(fn, ())
 
-    def test_named_tuple_redefine(self):
+    def test_namedtuple_redefine(self):
```
Nice job leaving the campground cleaner

gmagogsfm commented on test/test_jit.py:
```diff
@@ -119,6 +119,9 @@ def canonical(graph):
 def LSTMCellF(input, hx, cx, *params):
     return LSTMCell(input, (hx, cx), *params)
 
+class Point(NamedTuple):
```
I assume you define `Point` here because TorchScript can't compile class that's not defined in global scope? If so, you can actually define this class local to test case and use `make_global` to make TS see it.

gmagogsfm commented on torch/csrc/jit/python/python_sugared_value.cpp:
```diff
@@ -924,34 +924,22 @@ TypePtr registerNamedTuple(const py::object& obj, const SourceRange& loc) {
   TORCH_INTERNAL_ASSERT(isNamedTupleClass(obj));
   auto qualifiedName = c10::QualifiedName(py::cast<std::string>(
       py::module::import("torch._jit_internal").attr("_qualified_name")(obj)));
-  // Currently don't support default values
```
Love seeing deletions of code like this : )

gmagogsfm commented on test/test_jit.py:
```diff
@@ -119,6 +119,9 @@ def canonical(graph):
     return LSTMCell(input, (hx, cx), *params)
 
+class Point(NamedTuple):
+    x: Optional[torch.Tensor] = None
```
Nit: try some default value other than `None` as well for better test coverage

gmagogsfm reviewed (approved):
Looks good, approved with just a few minor comments.

Commit:
Update on "Support default values on NamedTuple fields"


Differential Revision: [D27327241](https://our.internmc.facebook.com/intern/diff/D27327241)

[ghstack-poisoned]

Commit:
Update on "Support default values on NamedTuple fields"


Differential Revision: [D27327241](https://our.internmc.facebook.com/intern/diff/D27327241)

[ghstack-poisoned]

Commit:
Update on "Support default values on NamedTuple fields"


Differential Revision: [D27327241](https://our.internmc.facebook.com/intern/diff/D27327241)

[ghstack-poisoned]

Commit:
Update on "Support default values on NamedTuple fields"


Differential Revision: [D27327241](https://our.internmc.facebook.com/intern/diff/D27327241)

[ghstack-poisoned]

gmagogsfm commented on test/test_jit.py:
```diff
@@ -14971,6 +15055,57 @@ def foo():
+        self.checkScript(fn, ())
+
+    def test_namedtuple_redefine(self):
+        global _1, _2
+        _1 = namedtuple('GoogLeNetOutputs', ['logits', 'aux_logits2', 'aux_logits1'])
+        _2 = namedtuple('GoogLeNetOutputs', ['different'])
```
Is this code copied from serialized model? The `_1` and `_2` seems a bit odd, please name them properly.

ansley replied:
Looks like this was a `rebase` issue. The original code is [here](https://github.com/pytorch/pytorch/blob/master/test/jit/test_typing.py#L535). I'll delete this

gmagogsfm commented on test/test_jit.py:
```diff
@@ -14971,6 +15055,57 @@ def foo():
+        _2 = namedtuple('GoogLeNetOutputs', ['different'])
+
+        with self.assertRaisesRegex(RuntimeError, r'redefine'):
+            @torch.jit.script
+            def foo(x, y):
+                # type: (_1, _2) -> _1
+                return x
```
Try to minimize amount of logic under `assertRaisesRegex` by defining `foo` outside but scripting it inside.

ansley replied:
Deleted (rebase issue, not my code)

gmagogsfm commented on test/test_jit.py:
```diff
@@ -13790,6 +13789,91 @@ def test_error():
+        #   graph(%x.1 : NamedTuple(x : Tensor, y : Tensor))
+        # It looks like similar test cases have had this issue as well
+        # (see: `test_namedtuple_python`).
+        FileCheck().check("NamedTuple")   \
+                   .check_next(r"return (%x.1)")    \
+                   .run(fn.graph)
```
What does the entire graph look like? I am wondering if there are better patterns to check than the word `NamedTuple` and `return`.

ansley replied:
```
graph(%x.1 : NamedTuple(x : Tensor, y : Tensor)):
  return (%x.1)
```

Maybe check for the whole first line? I try to avoid using variable names in `FileCheck` in the case that our naming logic changes later, but in this case it might be relevant

gmagogsfm commented on torch/_jit_internal.py:
```diff
@@ -975,19 +975,22 @@ def _try_get_dispatched_fn(fn):
 
 
-def _create_named_tuple(t, unqual_name: str, field_names: List[str]):
+def _create_named_tuple(t, unqual_name: str, field_names: Dict[str, Any]):
```
```suggestion
def _create_named_tuple(t, unqual_name: str, fields: Dict[str, Any]):
```

Should callsite of this method be updated too? 
https://github.com/pytorch/pytorch/blob/45012da298cdf5e37121080d9f4273fd8a58d070/torch/csrc/jit/python/pybind_utils.h#L727

ansley replied:
Thanks, I didn't know about this!

Commit:
Update on "Support default values on NamedTuple fields"


Differential Revision: [D27327241](https://our.internmc.facebook.com/intern/diff/D27327241)

[ghstack-poisoned]

ansley commented on torch/csrc/jit/serialization/import_source.cpp:
```diff
@@ -638,23 +638,27 @@ struct SourceImporterImpl : public Resolver,
+      auto name = Var(Assign(statement).lhs()).name().name();
+
+      fields.emplace_back(
+          std::pair<std::string, IValue>(std::move(name), IValue()));
```
Done, added tests

Commit:
Update on "Support default values on NamedTuple fields"


Differential Revision: [D27327241](https://our.internmc.facebook.com/intern/diff/D27327241)

[ghstack-poisoned]

Commit:
Update on "Support default values on NamedTuple fields"


Differential Revision: [D27327241](https://our.internmc.facebook.com/intern/diff/D27327241)

[ghstack-poisoned]

Commit:
Update on "Support default values on NamedTuple fields"


Differential Revision: [D27327241](https://our.internmc.facebook.com/intern/diff/D27327241)

[ghstack-poisoned]

Commit:
Update on "Support default values on NamedTuple fields"


Differential Revision: [D27327241](https://our.internmc.facebook.com/intern/diff/D27327241)

[ghstack-poisoned]

gmagogsfm commented on torch/csrc/jit/serialization/import_source.cpp:
```diff
@@ -485,13 +485,13 @@ struct SourceImporterImpl : public Resolver,
-          } else if (hook_names.find(def.name().name()) != hook_names.end()) {
-            hook_def_map.emplace(def.name().name(), def);
-            hook_resolver_map.emplace(def.name().name(), shared_from_this());
+          auto def_name = def.name().name();
+          if (pre_hook_names.find(def_name) != pre_hook_names.end()) {
+            pre_hook_def_map.emplace(def_name, def);
+            pre_hook_resolver_map.emplace(def_name, shared_from_this());
+          } else if (hook_names.find(def_name) != hook_names.end()) {
+            hook_def_map.emplace(def_name, def);
+            hook_resolver_map.emplace(def_name, shared_from_this());
```
This looks like a good clean up, but it is not related to this named tuple functionality in this PR, please make a separate PR for it.

ansley replied:
Sure, reverted

gmagogsfm commented on torch/csrc/jit/serialization/import_source.cpp:
```diff
@@ -639,23 +639,34 @@ struct SourceImporterImpl : public Resolver,
+      auto default_val = IValue();
+      if (assign.rhs().present()) {
+        std::vector<IValue> parsed = type_parser.evaluateDefaults(
+            assign.range(), {assign.rhs().get()}, {assign.type().get()});
```
What's the rationale behind using `assign.range()` instead of `assign.rhs().range()`?

ansley replied:
fixed

gmagogsfm commented on torch/csrc/jit/python/python_sugared_value.cpp:
```diff
@@ -930,40 +930,48 @@ TypePtr registerNamedTuple(const py::object& obj, const SourceRange& loc) {
+      auto type = tryToInferType(field_defaults[i]);
+      ival = toIValue(field_defaults[i], type.type());
+      TORCH_CHECK(
+          ival.tagKind() != "Tensor",
```
```suggestion
          !ival.isTensor(),
```

ansley replied:
Good, thanks!

gmagogsfm commented on torch/csrc/jit/python/python_sugared_value.cpp:
```diff
@@ -930,40 +930,48 @@ TypePtr registerNamedTuple(const py::object& obj, const SourceRange& loc) {
     TORCH_CHECK(
         type->isSubtypeOf(tt),
-        "Can't to redefine NamedTuple: ",
+        "Can't redefine NamedTuple after creation: ",
```
"after creation" may be a little confusing since it can refer to creation of named tuple instance.
```suggestion
        "Can't redefine NamedTuple: ",
```

ansley replied:
Sure, fixed

gmagogsfm replied:
Doesn't seem changed yet. Maybe because the latest revision is not pushed yet? In general, please push the fix before replying to the comment thread so that the change is visible and reviewer can confirm is the fix is what they meant.

gmagogsfm commented on torch/csrc/jit/python/python_sugared_value.cpp:
```diff
@@ -930,40 +930,48 @@ TypePtr registerNamedTuple(const py::object& obj, const SourceRange& loc) {
+      TORCH_CHECK(
+          ival.tagKind() != "Tensor",
+          "Tensors are not "
+          "supported as default NamedTuple fields");
```
Maybe state the reason here, we choose not to support it because it is dangerous to use a mutable object as default value.

gmagogsfm replied:
Something like this:
```
import torch
from typing import NamedTuple, List, Dict, Optional

t = torch.ones(1)

class Point(NamedTuple):
    x: Optional[List[int]] = None
    y: torch.Tensor = t

print(Point(1))
t[0] = 2
print(Point(1))
```

gmagogsfm commented on aten/src/ATen/core/type.cpp:
```diff
@@ -735,13 +742,28 @@ TupleTypePtr TupleType::createNamed(
         /*type=*/field_types[i],
-        /*N=*/i);
+        /*N=*/i,
+        /*default_value*/fields[i].second);
```
```suggestion
        /*default_value=*/fields[i].second);
```

ansley replied:
good, thanks

ansley commented on test/jit/test_list_dict.py:
```diff
@@ -1771,7 +1771,7 @@ def forward(self, x):
 
 class TestNamedTuple(JitTestCase):
-    def test_named_tuple(self):
+    def test_namedtuple(self):
```
Thanks! We've discussed this on some of the PRs that I've been working on concurrently. In the future, I'll always submit "clean up" PRs separately. This latest round of PRs should be the only ones that still have "clean up" logic in them. :)

ansley commented on torch/csrc/jit/frontend/script_type_parser.h:
```diff
@@ -30,7 +30,7 @@ class TORCH_API ScriptTypeParser {
   c10::IValue parseClassConstant(const Assign& assign);
 
- private:
+  // private:
```
Yes, thanks

ansley commented on torch/csrc/jit/python/pybind_utils.cpp:
```diff
@@ -67,7 +67,7 @@ IValue toIValue(py::handle obj, const TypePtr& type, c10::optional<int32_t> N) {
       py::tuple tuple = py::cast<py::tuple>(obj);
       size_t tuple_size = tuple.size();
-      auto tuple_type = type->cast<TupleType>();
+      TupleTypePtr tuple_type = type->cast<TupleType>();
```
okay

ansley commented on torch/csrc/jit/python/pybind_utils.h:
```diff
@@ -713,18 +713,39 @@ inline py::object toPyObject(IValue ivalue) {
+      std::vector<IValue> defaults;
+      auto it = std::find_if(
+          tuple_args.begin(), tuple_args.end(), [](const Argument& arg) {
+            if (arg.default_value()) {
+              return true;
+            }
+            return false;
```
Much better, thanks

ansley commented on torch/_jit_internal.py:
```diff
@@ -984,24 +984,29 @@ def _try_get_dispatched_fn(fn):
     # mypy: namedtuple() expects a string literal as the first argument
-    TupleType = collections.namedtuple(unqual_name, field_names)  # type: ignore[misc]
+    if sys.version_info < (3,7,0):
+        TupleType = collections.namedtuple(unqual_name, field_names)  # type: ignore[misc]
```
I needed to ignore two separate `mypy` errors that were on the same line. As far as I know, you can only specify "ignore all" (seen here as `type: ignore`) or "ignore one thing" (`type: ignore[COND]`). Please let me know if there's alternate syntax to ignore both, though.

The mypy errors come from: 1) `TupleType` being "redefined" on the next line (`mypy` doesn't recognize the conditional definition, and 2) the `defaults` field not existing in all Python versions (not an issue because of the conditional check)

ansley commented on torch/csrc/jit/python/pybind_utils.h:
```diff
@@ -713,18 +713,39 @@ inline py::object toPyObject(IValue ivalue) {
+            }
+            return false;
+          });
+      std::transform(
```
Where do you see that `std::transform` is only C++17/20? I thought that the parallelized version of `std::transform` was added in C++17, but `transform` itself has been around for a while. I might be mistaken, though

gmagogsfm replied:
You are right, I misremembered. Just checked cppreference and indeed `std::transform` is available.

ansley commented on torch/_jit_internal.py:
```diff
@@ -984,24 +984,29 @@ def _try_get_dispatched_fn(fn):
+def _create_named_tuple(t, unqual_name: str, field_names: List[str], defaults: List[Any]):
     # mypy: namedtuple() expects a string literal as the first argument
-    TupleType = collections.namedtuple(unqual_name, field_names)  # type: ignore[misc]
+    if sys.version_info < (3,7,0):
```
done

Commit:
Update on "Support default values on NamedTuple fields"


Differential Revision: [D27327241](https://our.internmc.facebook.com/intern/diff/D27327241)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27327241).

Commit:
Update on "Support default values on NamedTuple fields"


Differential Revision: [D27327241](https://our.internmc.facebook.com/intern/diff/D27327241)

[ghstack-poisoned]

Commit:
Update on "Support default values on NamedTuple fields"


Differential Revision: [D27327241](https://our.internmc.facebook.com/intern/diff/D27327241)

[ghstack-poisoned]

Commit:
Update on "Support default values on NamedTuple fields"


Differential Revision: [D27327241](https://our.internmc.facebook.com/intern/diff/D27327241)

[ghstack-poisoned]

Commit:
Update on "Support default values on NamedTuple fields"


Differential Revision: [D27327241](https://our.internmc.facebook.com/intern/diff/D27327241)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27327241).

Commit:
Update on "Support default values on NamedTuple fields"


Differential Revision: [D27327241](https://our.internmc.facebook.com/intern/diff/D27327241)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27327241).

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27327241).

gmagogsfm commented on aten/src/ATen/core/type.cpp:
```diff
@@ -742,13 +749,28 @@ TupleTypePtr TupleType::createNamed(
   std::vector<Argument> arguments;
-  for (size_t i = 0; i < field_names.size(); ++i) {
+  for (size_t i = 0; i < fields.size(); ++i) {
+    TORCH_CHECK(fields[i].second.tagKind() != "Tensor", "Tensors are "
```
Consider also state the reason for not supporting Tensor type default value here.

ansley replied:
Fixed

gmagogsfm commented on torch/_jit_internal.py:
```diff
@@ -993,24 +993,29 @@ def _try_get_dispatched_fn(fn):
     # mypy: namedtuple() expects a string literal as the first argument
-    TupleType = collections.namedtuple(unqual_name, field_names)  # type: ignore[misc]
+    if sys.version_info < (3, 7, 0):
+        TupleType = collections.namedtuple(unqual_name, field_names)  # type: ignore[no-redef, misc]
```
For Python before 3.7, I think there is a hacky way of creating a named tuple with default values:
https://stackoverflow.com/a/18348004

ansley replied:
Thanks!

gmagogsfm reviewed (approved):
Looks great, just a couple minor comment.

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27327241).

Commit:
Update on "Support default values on NamedTuple fields"


Differential Revision: [D27327241](https://our.internmc.facebook.com/intern/diff/D27327241)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27327241).

Commit:
Update on "Support default values on NamedTuple fields"


Differential Revision: [D27327241](https://our.internmc.facebook.com/intern/diff/D27327241)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27327241).

Commit:
Update on "Support default values on NamedTuple fields"


Differential Revision: [D27327241](https://our.internmc.facebook.com/intern/diff/D27327241)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27327241).

Commit:
Update on "Support default values on NamedTuple fields"


Differential Revision: [D27327241](https://our.internmc.facebook.com/intern/diff/D27327241)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27327241).

Commit:
Update on "Support default values on NamedTuple fields"


Differential Revision: [D27327241](https://our.internmc.facebook.com/intern/diff/D27327241)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27327241).

Commit:
Update on "Support default values on NamedTuple fields"


Differential Revision: [D27327241](https://our.internmc.facebook.com/intern/diff/D27327241)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27327241).

Commit:
Update on "Support default values on NamedTuple fields"


Differential Revision: [D27327241](https://our.internmc.facebook.com/intern/diff/D27327241)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27327241).

Commit:
Update on "Support default values on NamedTuple fields"


Differential Revision: [D27327241](https://our.internmc.facebook.com/intern/diff/D27327241)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27327241).

Commit:
Update on "Support default values on NamedTuple fields"


Differential Revision: [D27327241](https://our.internmc.facebook.com/intern/diff/D27327241)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27327241).

Commit:
Update on "Support default values on NamedTuple fields"


Differential Revision: [D27327241](https://our.internmc.facebook.com/intern/diff/D27327241)

[ghstack-poisoned]

Commit:
Update on "Support default values on NamedTuple fields"


Differential Revision: [D27327241](https://our.internmc.facebook.com/intern/diff/D27327241)

[ghstack-poisoned]

Commit:
Update on "Support default values on NamedTuple fields"


Differential Revision: [D27327241](https://our.internmc.facebook.com/intern/diff/D27327241)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27327241).

Commit:
Update on "Support default values on NamedTuple fields"


Differential Revision: [D27327241](https://our.internmc.facebook.com/intern/diff/D27327241)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27327241).

Commit:
Update on "Support default values on NamedTuple fields"


Differential Revision: [D27327241](https://our.internmc.facebook.com/intern/diff/D27327241)

[ghstack-poisoned]

Commit:
Update on "Support default values on NamedTuple fields"


Differential Revision: [D27327241](https://our.internmc.facebook.com/intern/diff/D27327241)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27327241).

Commit:
Update on "Support default values on NamedTuple fields"


Differential Revision: [D27327241](https://our.internmc.facebook.com/intern/diff/D27327241)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27327241).
